napari_aicssegmentation/view/workflow_steps_view.py

```diff
@@ -104,32 +104,21 @@ def _add_progress_bar(self):
                 labels_layout.addStretch()
         self.layout.addWidget(progress_labels)
 
-    def _add_workflow_steps(self, category: str):
-        # TODO: replace with real data
-        steps = [
-            {"number": 1, "name": "Intensity Normalization"},
-            {"number": 2, "name": "Edge Preserving Smoothing"},
-        ]
-
-        # Category label, e.g., "Preprocessing"
-        category_label = QLabel(category.upper())
+    def _add_workflow_steps(self, category: WorkflowStepCategory):
+        # Add category label, e.g., "Preprocessing"
+        category_label = QLabel(category.value.upper())
         category_label.setObjectName("categoryLabel")
         self.layout.addWidget(category_label)
 
-        # TODO: replace this with the data-driven WorkflowStep widget
+        # Gather all the steps in this category
+        steps = []
+        for step in self.all_steps:
```

Comment:
Alternative to loops for filtering a list: use [filter()](https://www.geeksforgeeks.org/filter-in-python/)
`filter` returns an iterator so the code would now look like this:
~~~python
steps = filter(lambda s: s.category == category, self.all_steps)
for step in steps:
   self.layout.addWidget(WorkflowStepWidget(step))
~~~
Or as a one-liner:
~~~python
for step in filter(lambda s: s.category == category, self.all_steps)
   self.layout.addWidget(WorkflowStepWidget(step))
~~~

Reply:
Thanks! I forgot that existed, and I must've had something else going on here at some point with `enumerate` here..... 🤔